src/main/kotlin/com/parmet/buf/gradle/GenerateTask.kt

```diff
@@ -26,10 +26,43 @@ abstract class GenerateTask : DefaultTask() {
         execBuf(args + additionalArgs())
     }
 
-    private fun additionalArgs() =
-        if (getExtension().generateOptions?.includeImports == true) {
+    private fun additionalArgs(): List<String> {
+        val generateOptions = getExtension().generateOptions
+        val importOptions = if (generateOptions?.includeImports == true) {
             listOf("--include-imports")
         } else {
             emptyList()
         }
+
+        val configOptions = genConfigFile()?.let {
+            listOf("--template", it.absolutePath)
+        } ?: emptyList()
+
+        return importOptions + configOptions
+    }
+
+    private fun genConfigFile(): File? {
+        return getExtension().generateOptions?.let { generateOptions ->
+            val selectedTemplateFiles = listOfNotNull(generateOptions.templateFileLocation, project.file("buf.gen.yaml")).filter {
+                it.exists() && it.isFile
+            }
+            check(selectedTemplateFiles.isNotEmpty()) {
+                if (generateOptions.templateFileLocation != null) {
+                    "Specified templateFileLocation does not exist."
+                } else {
+                    "No buf.gen.yaml file found in the root directory."
+                }
+            }
+            check(selectedTemplateFiles.size == 1) {
+                "Buf gen template file specified in the root directory as well as with templateFileLocation; pick one."
+            }
+
+            val templateFile = selectedTemplateFiles.single()
+            check(generateOptions.templateFileLocation == null || templateFile == generateOptions.templateFileLocation) {
```

Comment:
Is this check still needed?

Reply:
This will go away with the logic refactoring below